### PR TITLE
Constraints for flags configuration

### PR DESCRIPTION
**Pull request by querolita**
These constraints were not in the Kimchi gate because flags were publicly accessible as gate coefficients encoding the mode of operation.

### CODE REVIEW

File: optimism/src/keccak/constraints.rs [hunk 1]
```diff
@@ -26,8 +30,38 @@ impl<Fp: Field> Constraints for KeccakEnv<Fp> {
...
     fn constraints(&mut self) {
-        todo!();
         // CORRECTNESS OF FLAGS
+        {
+            // TODO: remove redundancy if any
+
+            // Booleanity of sponge flags
+            {
+                // Absorb is either true or false
+                self.constrain(Self::boolean(self.absorb()));
```
Comment (dannywillems) [inline optimism/src/keccak/constraints.rs:40]: Follow-up: I think the operations defined in BoolOps overlap with some from the interpreter env.

Comment (dannywillems) [reply optimism/src/keccak/constraints.rs:40]: Opened https://github.com/o1-labs/proof-systems/issues/1609

File: optimism/src/keccak/constraints.rs [hunk 2]
```diff
@@ -26,8 +30,38 @@ impl<Fp: Field> Constraints for KeccakEnv<Fp> {
...
+                // Squeeze is either true or false
+                self.constrain(Self::boolean(self.squeeze()));
```
Comment (dannywillems) [inline optimism/src/keccak/constraints.rs:42]: Nit: I would have called the method `is_boolean`. Not part of this PR.

File: optimism/src/keccak/constraints.rs [hunk 3]
```diff
@@ -26,8 +30,38 @@ impl<Fp: Field> Constraints for KeccakEnv<Fp> {
...
+                // Root is either true or false
+                self.constrain(Self::boolean(self.root()));
+                // Pad is either true or false
+                self.constrain(Self::boolean(self.pad()));
+            }
+            // Mutually exclusiveness of flags
+            {
+                // Squeeze and Root are not both true
+                self.constrain(Self::either_false(self.squeeze(), self.root()));
```
Comment (dannywillems) [inline optimism/src/keccak/constraints.rs:51]: What's the expected output of `either_false`?
If at least one of the parameters is `false` (i.e. `0`), return `true` (i.e. `1`)?

Comment (querolita) [reply optimism/src/keccak/constraints.rs:51]: if at least one of them is false, then the constrain is satisfied -> the constrain is not satisfied if both are true -> you cannot have an absorb root in a squeeze sponge

File: optimism/src/keccak/constraints.rs (hunk 2 above)
Comment (querolita) [reply optimism/src/keccak/constraints.rs:42]: oh, yeah I can modify this. I just inherited a function we had in kimchi that was called just `boolean`

File: optimism/src/keccak/constraints.rs (hunk 1 above)
Comment (querolita) [reply optimism/src/keccak/constraints.rs:40]: Probably, they were created in parallel

**Review by dannywillems — APPROVED** [review]
